config.go

```diff
@@ -0,0 +1,152 @@
+package main
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"log"
+	"net/url"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/tailscale/hujson"
+)
+
+// configFile represents the on-disk configuration for this tsproxy instance.
+type config struct {
+	// StateDir is where app state is stored, this should be persisted between
+	// rounds. Defaults to the user cache dir. If the Kubernetes config is set,
+	// it overrides this.
+	StateDir string `json:"stateDir"`
+	// Kubernetes configures the proxy to run in a kubernetes cluster. In this
+	// case the StateDir is ignored, and state managed in a secret.
+	Kubernetes kubernetesConfig `json:"kubernetes"`
+	// MetricsDiscovery port sets the port we should listen for internal items
+	// for, i.e metrics and discovery info
+	MetricsDiscoveryPort int `json:"port"`
+	// LogTailscale indicates that we should log tailscale output
+	LogTailscale bool       `json:"logTailscale"`
+	Upstreams    []upstream `json:"upstreams"`
+}
+
+// ConfigUpstream represents the configuration for a single upstream for this
+// tsproxy instance.
+type upstream struct {
+	// Name for this upstream. This is what it will be registered in tailscale
+	// as.
+	Name string `json:"name"`
+	// Backend is the URL to the backend that serves this upstream
+	Backend    string `json:"backend"`
+	Prometheus bool   `json:"prometheus"`
+	Funnel     bool   `json:"funnel"`
+
+	// FunnelPublicPatterns sets the list of patterns where public (aka
+	// unauthenticated) access are allowed. If not set, no open access is
+	// permitted - and auth must be configured. The patterns are in
+	// http.ServeMux format.
+	FunnelPublicPatterns []string `json:"funnelPublicPatterns"`
+
+	// OIDCIssuer sets the issuer to authenticate funnel access with. Any
+	// patterns not labeled as public will be handled by this.
+	OIDCIssuer string `json:"oidcIssuer"`
+	// OIDCClientID sets the OIDC client ID
+	OIDCClientID string `json:"oidcClientID"`
+	// OIDCClientSecret sets the OIDC client secret
+	OIDCClientSecret string `json:"oidcClientSecret"`
```

Comment:
be cool to support a file too, play nice with systemd-creds. For flags I usually do this:

```
	if _, path, ok := strings.Cut(*cliSecret, "file://"); ok {
		b, err := os.ReadFile(path)
		if err != nil {
			return err
		}
		s := strings.TrimSpace(string(b))
		cliSecret = &s
	}
```

but doesn't really work w/ a config file. Maybe oidcClientSecretFile?

Reply:
also wondering if this should be on the config struct, or do you plan on using a different issuer for some upstream(s)?